proposals/0095-xof-hashing.md

```diff
@@ -0,0 +1,325 @@
+, or exchanging the `Keccak` implemenation with a `cSHAKE` implementation.
+
+### STROBE
+
+Strobe is a protocol framework based on the duplex Sponge construction. 
+`Strobe-f-λ/b` is a Strobe instance with a targeted security level of `λ` bits. 
+The capacity `c = 2λ`, the bandwidth `b=r+c`, where `r` is the rate, and 
+`F=f(b)` is the sponge function, based on Keccak-f. 
+
+Strobe can be instantiated with cSHAKE and specifies two instances based on 
+cSHAKE:
+
+- `Strobe-128/1600()`
+- `Strobe-256/1600()`
+
+where the  initial state of `Strobe-f-λ/b` is
+
+```
+  S_0 := F(bytepad(encode_string("") || encode_string("STROBEv1.0.2"), r/8))
+       = F([[1, r/8, 1, 0, 1, 96]] || "STROBEv1.0.2" || (r/8-18) * [[0]])
+```
+
+For `Strobe-128/1600()` any data squeezed is of the form `cSHAKE128(X)` and for 
+`Strobe-256/1600()` it is of the form `cSHAKE256(X)`. 
+
+#### Implementation Details
+
+The Strobe designers released an official implementation in C available at 
+[https://sourceforge.net/p/strobe](https://sourceforge.net/p/strobe/code/ci/master/tree/)
+. Moreover, a minimal Strobe-128 implementation in Rust is 
+available in the 
+[source code](https://github.com/zkcrypto/merlin/blob/main/src/strobe.rs) for 
+the merlin crate. 
+
+The STROBE protocol framework can be build on top of the current Keccak 
+implementation or on top of the cSHAKE implementation, as outlined in the 
+Strobe-128 
+[implementation](https://github.com/zkcrypto/merlin/blob/main/src/strobe.rs) 
+of the merlin crate. Additionally to hashing, that is currently 
+implemented for Keccak, additional sponge functions need to be implemented, 
+such as: 
+
+- Adding associated data 
+- Key Addition
+- Extract hash/pseudorandom data (PRF)
+
+which should all be available in Keccak/cSHAKE, by using the `Absorb` and 
+`Squeeze` functions. In more details, for adding associated data the functions 
+`AD` and `meta_AD` need to be implemented, that absorbs data into the state. 
+`meta_AD` describes the protocols interpretation of the operation. The function 
+`KEY` adds a cryptographic key to the state by absorbing the key. The function 
+`PRF` extracts pseudorandom data from the state, by squeezing data. 
+Additionally, Strobe operations are defined by flags as outlined in the 
+specifications at 
+[https://strobe.sourceforge.io/specs/#ops.flags](https://strobe.sourceforge.io/specs/#ops.flags).
+
+For all the above functions, syscalls need to be defined in 
+[solana/sdk/program/src/syscalls/definition.rs](https://github.com/solana-labs/solana/blob/master/sdk/program/src/syscalls/definitions.rs#L46)
+ as follows:
+
+```
+define_syscall!(fn sol_strobe128_ad(...) -> u64);
+define_syscall!(fn sol_strobe128_meta_ad(...) -> u64);
+define_syscall!(fn sol_strobe128_key(...) -> u64);
+define_syscall!(fn sol_strobe128_prf(...) -> u64);
+
+define_syscall!(fn sol_strobe256_ad(...) -> u64);
+define_syscall!(fn sol_strobe256_meta_ad(...) -> u64);
+define_syscall!(fn sol_strobe256_key(...) -> u64);
```

Comment:
Every new syscall adds overhead and hurts performance for the VM vs an SBPF implementation for some specific app. I would recommend finding an approach similar to what @ripatel-fd has mentioned.